pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+	FinalizerName         = "instance.clusterlink.net/finalizer"
+	ExternalNodePort      = 30443
+	StatusModeReady       = "Ready"
+	StatusModeProgressing = "ProgressingMode"
+)
+
+// InstanceReconciler reconciles a ClusterLink instance object
+type InstanceReconciler struct {
+	client.Client
+	Scheme      *runtime.Scheme
+	instance    *api.Instance
+	clNamespace string
+	Logger      *logrus.Entry
+	CaFabric    []byte
+}
+
+// +kubebuilder:rbac:groups=clusterlink.net,resources=instances,verbs=list;get;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=clusterlink.net,resources=instances/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=clusterlink.net,resources=instances/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=list;get;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=services;namespaces,verbs=list;get;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list;get;watch
+// +kubebuilder:rbac:groups="apps",resources=deployments,verbs=list;get;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=clusterroles;clusterrolebindings,verbs=list;get;watch;create;update;patch;delete
+// TODO- should review the operator RABCs.
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *InstanceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&api.Instance{}).
+		Watches(
+			&appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: ControlPlaneName,
+				},
+			},
+			&handler.EnqueueRequestForObject{},
+		).
+		Watches(
+			&appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: DataPlaneName,
+				},
+			},
+			&handler.EnqueueRequestForObject{},
+		).
+		Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// The reconcile get instance YAML and creates the ClusterLink components.
+func (r *InstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// Fetch the Clusterlink instance
+	instance := &api.Instance{}
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Requeue the request on error
+		return ctrl.Result{}, err
+	}
+
+	// CRD details
+	r.Logger.Infof("Enter instance Reconcile - (Namespace: %s, Name: %s)", instance.Namespace, instance.Name)
+	r.Logger.Info("InstanceSpec- ",
+		" DataPlane.Type: ", instance.Spec.DataPlane.Type,
+		", DataPlane.Replicas: ", instance.Spec.DataPlane.Replicas,
+		", Ingress.Type: ", instance.Spec.Ingress.Type,
+		", LogLevel: ", instance.Spec.LogLevel,
+		", ContainerRegistry: ", instance.Spec.ContainerRegistry,
+		", ImageTag: ", instance.Spec.ImageTag,
+	)
+	r.instance = instance
```

Comment:
It might be sufficient to just store a NamespacedName here, since you should have access to the Instance object and retrieve its fields on demand (it'll be cached in the controller-runtime). Having the instance namespace and name is also sufficient to mark additional instances as in conflict and provide the original object name.
I'm not a fan of copying and keeping a full copy on the reconciler.

Reply:
This copy is only used in a single call of the reconciliation, I use this copy in part of the functions and also update the status,
I can change it to a variable and pass it to all the functions, but I prefer this way.